python/pyspark/sql/utils.py

```diff
@@ -24,7 +28,11 @@ def __init__(self, desc, stackTrace):
         self.stackTrace = stackTrace
 
     def __str__(self):
-        return repr(self.desc)
+        desc = self.desc
+        if isinstance(desc, unicode):
+            return str(desc.encode('utf-8'))
```

Comment:
@ueshin, you are right and I misread the codes. We need to

- unicode in Python 2 => `u.encode("utf-8")`.
- others in Python 2 => return `str(s)`.
- others in Python 3 => return `str(s)`.

Root cause for https://github.com/apache/spark/pull/17267#issuecomment-308231375 looks because `encode` on string (also same as unicode in Python 2) in Python 3 produces 8-bit bytes, `b"..."`, (also same as normal string, `"..."` and `b"..."`, where `b` is ignored, in Python 2). And `str` function works differently as below:

Python 2

```python
>>> str(b"aa")
'aa'
>>> b"aa"
'aa'
```

Python 3

```python
>>> str(b"aa")
"b'aa'"
>>> "aa"
'aa'
```